plugins/chat/assets/javascripts/discourse/components/chat-composer.js

```diff
@@ -1,172 +1,249 @@
-export default Component.extend(TextareaTextManipulation, {
-  chatChannel: null,
-  chat: service(),
-  classNames: ["chat-composer-container"],
-  classNameBindings: ["emojiPickerVisible:with-emoji-picker"],
-  chatEmojiReactionStore: service("chat-emoji-reaction-store"),
-  chatEmojiPickerManager: service("chat-emoji-picker-manager"),
-  chatStateManager: service("chat-state-manager"),
-  timer: null,
-  value: "",
-  inProgressUploads: null,
-  composerEventPrefix: "chat",
-  composerFocusSelector: ".chat-composer-input",
-  canAttachUploads: reads("siteSettings.chat_allow_uploads"),
-  isNetworkUnreliable: reads("chat.isNetworkUnreliable"),
-  typingMention: false,
-  chatComposerWarningsTracker: service(),
-
-  @discourseComputed(...chatComposerButtonsDependentKeys())
-  inlineButtons() {
-    return chatComposerButtons(this, "inline", this.context);
-  },
-
-  @discourseComputed(...chatComposerButtonsDependentKeys())
-  dropdownButtons() {
-    return chatComposerButtons(this, "dropdown", this.context);
-  },
-
-  @discourseComputed("chatEmojiPickerManager.{opened,context}")
-  emojiPickerVisible(picker) {
-    return picker.opened && picker.context === "chat-composer";
-  },
-
-  @discourseComputed("chatStateManager.isFullPageActive")
-  fileUploadElementId(fullPage) {
-    return fullPage ? "chat-full-page-uploader" : "chat-widget-uploader";
-  },
-
-  init() {
-    this._super(...arguments);
-
-    this.appEvents.on(
-      "upload-mixin:chat-composer-uploader:in-progress-uploads",
-      this,
-      "_inProgressUploadsChanged"
+import { isPresent } from "@ember/utils";
+
+export default class ChatComposer extends Component {
+  @service capabilities;
+  @service site;
+  @service siteSettings;
+  @service chat;
+  @service chatComposerPresenceManager;
+  @service chatComposerWarningsTracker;
+  @service appEvents;
+  @service chatEmojiReactionStore;
+  @service chatEmojiPickerManager;
+
+  @tracked isFocused = false;
+
+  get pane() {
+    return this.args.paneService;
+  }
+
+  get composer() {
+    return this.args.composerService;
+  }
+
+  get inlineButtons() {
+    return chatComposerButtons(this, "inline", this.args.context);
+  }
+
+  get dropdownButtons() {
+    return chatComposerButtons(this, "dropdown", this.args.context);
+  }
+
+  get fileUploadElementId() {
+    return this.args.context + "-file-uploader";
+  }
+
+  get canAttachUploads() {
+    return (
+      this.siteSettings.chat_allow_uploads &&
+      isPresent(this.args.uploadDropZone)
     );
+  }
 
-    this.setProperties({
-      inProgressUploads: [],
-      _uploads: [],
-    });
+  @action
+  setupAutocomplete(textarea) {
+    const $textarea = $(textarea);
+    this.#applyUserAutocomplete($textarea);
+    this.#applyEmojiAutocomplete($textarea);
+    this.#applyCategoryHashtagAutocomplete($textarea);
+  }
 
-    this.composerService?.registerFocusHandler(() => {
-      this._focusTextArea();
-    });
-  },
+  @action
+  setupTextareaInteractor(textarea) {
```

Comment:
This is much nicer than having to inherit from `TextareaTextManipulation` :+1: I really want to change how my uppy code is at some point to do this insted of an inheriting chain. Maybe David will do it for me if I wait long enough :P

Reply:
Yes I will need to talk with him about some of my solutions in this PR, see if there's something better to do or patterns to extract in our codebase.